apps/safety/main.cpp

```diff
@@ -1,13 +1,219 @@
 
 int main()
 {
+    // setup bus
+    i2c.frequency(10000);
+    initCAN();
+    // turn led off, turn on if issue occurs
+    led1 = 0;
+    led2 = 0;
+    led3 = 0;
+
     while (true) {
-        led1 = !led1;
-        wait_ms(k_interval_ms);
+        //read can messages to get data to configure device
+        u_int8_t command = 0;
+        if(can.read(rxMsg))
+        {
+            processCANMsg(&rxMsg);
+            rxMsg >> command;
+            rxMsg.clear();
+        }
+
+        float measuredCurrent = getCurrentData();
+        if (inRange(currentUpperLim, currentLowerLim, measuredCurrent))
+        {
+            led1 = 1;
+        }
+
+        float measuredVoltage = getVoltageData();
+        if (inRange(voltageUpperLim, voltageLowerLim, measuredVoltage))
+        {
+            led2 = 1;
+        }
+
+        float measuredTemp = readThermosistor();
+        if(inRange(tempUpperLim, tempLowerLim, measuredTemp))
+        {
+            led3 = 1;
+        }
     }
 }
+
+void initCAN()
+{
+    // will need new id's 
+    can.filter(CAN1_RX, CAN1_TX, ROVER_CANID_MASK, CANStandard);
+}
+
+void processCANMsg(CANMsg *p_newMsg)
+{
+    // PRINT_INFO("Recieved CAN message with ID %X\r\n", p_newMsg->id);
+    // The specific can ID for configuring chip is ___
+    switch (p_newMsg->id){
+        case expectedCANID:
+            pc.printf("Updating INA226 config\r\n");
+            break;
+        default:
+            pc.printf("Recieved unimplemented command\r\n");
+            break;
+    }
+}
+
+
+int inRange(float upperLim, float lowerLim, float val)
+{
+    int inRange = (val > upperLim || val< lowerLim) ? 1: 0;
+    return inRange;
+}
+
+
+float getCurrentData()
+{
+    //some conversion factor for current data
+    float conversionFactor = 1;
+    //may need char16_t?
+    char cmd[3] = {CURRENT_REGISTER};
+    
+    i2c.write(SENSOR_SLAVE_ADDRESS, cmd, 2);
+    i2c.read(SENSOR_SLAVE_ADDRESS, cmd, 2, false);
+    
+    char16_t currentData = cmd[1]; 
+    currentData = (cmd[2] << 8 | 0x00);
+    
+    return currentData * conversionFactor;
+}
+
+
+float getVoltageData()
+{
+    //some conversion factor for voltage data
+    float conversionFactor = 1;
+    char cmd[3] = {VOLTAGE_REGISTER};
+    
+    i2c.write(SENSOR_SLAVE_ADDRESS, cmd, 2);
+    i2c.read(SENSOR_SLAVE_ADDRESS, cmd, 2, false);
+    
+    char16_t voltageData = cmd[1]; 
+    voltageData = (cmd[2] << 8 | 0x00);
+    
+    return voltageData * conversionFactor;
+}
+
+
+float readThermosistor(float vin, float vout)
+{
+    //voltage divider
+    float rthermo = rResistor *((vin/vout)-1);
+
+    //convert resistance to temp (steinhart and hart eqn)
+    return 1/(A + B * log(rthermo) + C * pow(log(rthermo),3));
+}
+
+int configureSensor(configurationMode operationMode, int resetRegisters)
+{
+    char cmd[3] = {CONFIG_REGISTER, 0x00, 0x00};
+    int dataByte = 0x00;
+
+    //fill up register with configuration data
+    switch (operationMode)
```

Comment:
this switch statement seems redundant, it's no different from dataByte |= powerReset